data/finset.lean

```diff
@@ -623,6 +625,66 @@ finset.induction_on s ⟨0, empty_subset _⟩ $ λ a s ha ⟨n, hn⟩,
 
 end range
 
+/- Ico (a closed open interval) -/
+variables {n m l : ℕ}
+
+/-- `Ico n m` is the set of natural numbers `n ≤ k < m`. -/
+def Ico (n m : ℕ) : finset ℕ := ⟨_, nodup_Ico n m⟩
+
+namespace Ico
+
+@[simp] theorem val (n m : ℕ) : (Ico n m).1 = multiset.Ico n m := rfl
+
+@[simp] theorem mem : l ∈ Ico n m ↔ n ≤ l ∧ l < m := multiset.Ico.mem
+
+@[simp] theorem Ico_range (n : ℕ) : Ico 0 n = range n :=
+by dsimp [Ico]; congr; dsimp [multiset.Ico, list.Ico, multiset.range]; rw list.range_eq_range'
+
+@[simp] theorem empty (n : ℕ) : Ico n n = ∅ :=
+by dsimp [Ico]; congr; simp
+
+@[simp] theorem succ_singleton (n : ℕ) : Ico n (n+1) = singleton n :=
+by dsimp [Ico]; congr; simp
+
+lemma pred_singleton {m : ℕ} (h : m > 0) : Ico (m-1) m = {m-1} :=
+by dsimp [Ico]; congr; rw multiset.Ico.pred_singleton h; refl
+
+@[simp] lemma filter_lt (l n m : ℕ) : (Ico n m).filter (λ x, x < l) = (Ico n (min m l)) :=
+begin
+  /- `tidy` says -/
+  ext,
+  simp,
+  fsplit,
+  intros a_1, cases a_1, cases a_1_left, fsplit, assumption,
+  swap,
+  intros a_1, cases a_1, fsplit, fsplit, assumption,
+  swap 3,
+  -- and after that it's just chasing down inequalities...
+  apply lt_min; assumption,
+  apply lt_of_lt_of_le a_1_right,
+  apply min_le_left,
+  apply lt_of_lt_of_le a_1_right,
+  apply min_le_right,
+end
+
+@[simp] lemma diff (l n m : ℕ) : (Ico n m) \ (Ico n l) = Ico (max n l) m :=
```

Comment:
This lemma should probably have a sibling
```
@[simp] lemma diff' (l n m : ℕ) : (Ico n m) \ (Ico l m) = Ico n (min m l) :=
```

Reply:
Yes (and better names for both).

Reply:
And similarly we should have
`lemma filter_ge`.

Reply:
`diff_right` and `diff_left` perhaps?

Reply:
Sounds good to me

Reply:
Okay, I've put in stubs for these.